File: workspace/2023/2023-05-13/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fn n -- fn )
    dup 0 = if
        drop        ( fn n -- fn )
    else
        dup rot *   ( fn n -- n n*fn )
        swap 1 -    ( n n*fn -- n*fn n-1 )
        _factorial  ( n*fn n-1 -- fn )
    endif
;

: factorial ( n -- fn )
    1 swap      ( n -- 1 n )
    _factorial  ( 1 n -- fn )
;

: _show_factorials recursive ( max min -- )
    over over < if
        drop drop
    else
        dup factorial .
        1 +
        _show_factorials
    endif
;

: show_factorials ( n -- )
    s\" factorials:\n" type
    1 _show_factorials
    s\" \n" type
;

File: workspace/2023/2023-03-18/numerical-sequence.prolog/factorial.pl

%
% N 番目の階乗数.
%

factorial(0, 1).

factorial(N, Fn) :-
    N > 0,
    is(M, N - 1),
    factorial(M, Fm),
    is(Fn, N * Fm).

%
% Min 番目から Max 番目までの階乗数.
%

factorials(Min, Max, []) :-
    Min > Max.

factorials(Min, Max, [Fmin|Xs]) :-
    Min =< Max,
    factorial(Min, Fmin),
    is(NextMin, Min + 1),
    factorials(NextMin, Max, Xs).

